### ISSUE TITLE
Messages sent while offline disappear after connection is regained if the user closes/reopen app - Reported by: @Santhosh-Sellavel 

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#5987]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Make sure you are offline.
2. Send some messages to any user.
3. Refresh the page on the web. Close & open the app on mobile.
4. Now you can still the messages.
5. Now enable internet
6. Repeat step 3 or switch between chats.

## Expected Result:
Messages that were sent while offline should be sent to the other user 

## Actual Result:
Messages are displayed on chat but disappear once the user regains the connection. 

## Workaround:
None, sent messages while offline are lost. 

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.8-8

**Reproducible in staging?:** Yes 
**Reproducible in production?:** Yes

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/44479856/138369051-b93909f6-2e1f-4ab7-b7e9-c81d71dd0f45.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** @Santhosh-Sellavel 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1634765287021500

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
@johnmlee101 6 days overdue. This is scarier than being forced to listen to Vogon poetry!

**Comment by MelvinBot** [comment]
@johnmlee101 10 days overdue. I'm getting more depressed than Marvin.

**Comment by MelvinBot** [comment]
@johnmlee101 12 days overdue. Walking. Toward. The. Light...

**Comment by kidroca** [comment]
"Add Comment" requests are persisted to storage when we're offline
But when we regain connectivity they are read, removed from storage and kept only in memory
Making a refresh/exit before they are complete would make the app "forget" about them

#### 1. Persisted request getting cleared when they're added to the in-memory queue
https://github.com/Expensify/App/blob/b12eb5f9f18ba6578790b55ad248b682724f77ae/src/libs/Network.js#L55-L59

#### 2. Persisted requests getting cleared by the queue itself
https://github.com/Expensify/App/blob/b12eb5f9f18ba6578790b55ad248b682724f77ae/src/libs/Network.js#L215-L220

The request are only moved in memory, they are initiated but not completed  - making a browser refresh (or closing the app) after the above code has emptied persistent storage would lose the data when the request didn't complete

**Comment by kidroca** [comment]
In theory it should be possible to recreate this with shorter steps like
1. Make sure you are offline.
2. Send some messages to any user.
3. Regain connectivity
4. Very shortly after refresh or close the page / quit the app
When you open the app again you should see the messages briefly and then they would disappear (These were the optimistic messages created and stored. They get cleared since we receive the latest data from the server and it is supposed to replace them with the actual messages, but since requests never made it they're just removed)

Considering that it's very specific window of time that will cause the issue it's less likely to happen often

IRL cases that this bug can happen are:
- User have been offline for an extended period of time and has accumulated many requests. They decide to quit/terminate the app, but by coincidence they regain connectivity right before they quit
- User with accumulated requests regains connectivity. Something causes the app to refresh - like an app update
- User with accumulated requests regains connectivity. Something causes the app to crash
- User with accumulated requests regains connectivity. Something causes a logout

All of these are the case of: _A persisted request is moved to memory and executed but something unexpected happens_

---

IMO we're only trying to "bend" the network queue to fit this "process persisted requests" functionality
Due to the cyclic nature of the default queue and that it has no context of what's already running we're forced to clear persisted requests otherwise they'll get repeated - ` Onyx.set(ONYXKEYS.NETWORK_REQUEST_QUEUE, [])`
Instead a separate queue should be processing persisted requests
- it would not be interval based - it's only needed when we have persisted requests - it's triggered when we regain connectivity
- it would clear requests from storage as they complete, and not when they are moved to the queue
- the default queue is relieved

**Comment by johnmlee101** [comment]
Looking now!

**Comment by johnmlee101** [comment]
Hmm, I would also say that a separate queue should have a maximum timeout, so we should keep that data with the requests. However, I like the idea. Do you know which files would need to be modified, or some psuedo code to represent this change?

**Comment by kidroca** [comment]
I think the only change to acomodate the suggestion would be in https://github.com/Expensify/App/blob/main/src/libs/Network.js

1. We still use the general queue to persist `retryableRequests` when we're offline
2. We remove any other handling related to persisted requests, this code
3. When we're back online we process whatever is stored in persistent request storage
```js
Onyx.connect({
    key: ONYXKEYS.NETWORK,
    callback: (network) => {
        // when we transition from offline to online trigger the persisted request queue
        if (isOffline && !network.isOffline) {
            startPersistedRequestsQueue();
        }

        isOffline = network.isOffline;
    }
});

function startPersistedRequestsQueue() {
  // If the queue is already running we don't have to do anything
  // it will sort itself in case more requests are persisted after it started
  if (persistedRequestsQueueTask) {
     return;
  }

  processPersistedRequestsQueue();
}

function processPersistedRequestsQueue() { 
  // Onyx.get is necessary in the event that persisted requests are not yet read when we call this function
  // I'll mention alternatives outside the example
  persistedRequestsQueueTask = Onyx.get(ONYXKEYS.NETWORK_REQUEST_QUEUE)
      .then(persistedRequests => {
          // This sanity check is also a recursion exit point
          if (_.size(persistedRequests) === 0 || isOffline) {
              persistedRequestsQueueTask = null;
              return;
          }

          const tasks = _.map(persistedRequests, request => {
              // processRequest - reusable logic common to both request queues (should return a promise)
              return processRequest(request)
                  .then(() => removeFromPersistedStorage(request))  // when a request completes remove it from storage
                  .catch((error) => {
                       // Decide whether to discard the request despite the error, or keep it for a retry
                       if (!shouldRetryRequest(error)) {
                           return removeFromPersistedStorage(request);
                       }
                  });
            });

            // More request could have been persisted by now
            // If that's the case we'll handle them with a recursive call
            // Otherwise the recursion will exit at the recursion exit point
            return Promise.allSettled(tasks)
                .finally(roessPersistedRequestsQueue)
    })
}

// Have something unique in every request and use it to distinct which one we ought to remove
function removeFromPersistedStorage(request) {
  return Onyx.get(ONYXKEYS.NETWORK_REQUEST_QUEUE)
        .then(persistedRequests => {
           const nextPersistedRequests = _.filter(persistedRequests, r => r.id !== request.id);
           return Onyx.set(ONYXKEYS.NETWORK_REQUEST_QUEUE, nextPersistedRequests);
        });
}
```

Additional requirements like a request timeout can be implemented in `processRequest` though it might not be necessary - request would timeout on their own 
If for some reason we need to abort the queue from outside, additional APIs can be exposed

---

This queue would start as much request as possible in parallel, the underlying environment would prioritize how they're actually executed, but we can assume a pool of 5-10 requests running at the same time

We might be fine with a simpler queue that processes one request at a time, then removes it from storage and reads the next one and so on until we run out of processed requests.
This way we can also order and execute requests based on a timestamp

---

### Alternatives to `Onyx.get` 

#### Currently applied workaround - connect and disconnect
https://github.com/Expensify/App/blob/ccdf7293a67d07df593f79a6792c20871c7b8fe6/src/libs/Network.js#L77-L83
- we lose the ability to promise chain and we'll have to deal with flags for the recursive part of the queue

#### Top level Onyx.connect call for everything that we need
Not as reliable as the Onyx.get version - we're basically hoping that everything we need is already updated by the top level connect when we use it
Here we're discussion an alternative to top level connections: https://github.com/Expensify/App/issues/6151

**Comment by kidroca** [comment]
Hmm seems like I misunderstood the timeout part, you mean like persisted requests getting stale after a day or so? 

We can include some stale request filtering logic in there

**Comment by johnmlee101** [comment]
Ah yeah sorry, if things become too stale we should ignore. I'd imagine there's a situation where you're offline with a backlog of messages and you only check days weeks or months from then and having it auto-post wouldn't be useful. 

So yep! Stale request filtering.

**Comment by johnmlee101** [comment]
I like this though! I'm going to mark as external, get a manager to create the job then I'll have them assign you.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @stephanieelliott (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by johnmlee101** [comment]
Sorry for the shuffle there. @stephanieelliott once you create the job can you assign kidroca?

**Comment by kidroca** [comment]
`Onyx.get` fate is still undecided - I've posted alternatives to highlight their problems - none are suitable 
Until this problem is solved: https://github.com/Expensify/App/issues/6151 I don't think my suggestion can be aplied

**Comment by MelvinBot** [comment]
@johnmlee101, @stephanieelliott Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by johnmlee101** [comment]
Still on hold, I presume?

**Comment by kidroca** [comment]
#6151 was closed with the decision that we don't have sufficient cases that we need to do something about it (like exposing `Onyx.get`) just yet

Maybe we can come up an updated proposal that doesn't need `Onyx.get`.
This is a general overview of the strategy
1. When we're offline some requests (`shouldPersist: true`) are stored to disk by the main queue
2. When we transition from offline to online we read what's stored and send it
      - we should also do this at startup
3. When each item complets we're free to remove it from storage
4. When the queue is already running we want to avoid triggering it again 

1 and 2 are the easy parts
3 - This is a key functionality - keeping **each** request on disk until it's completed. In comparison we can wait for all the request and then delete the whole offline queue from disk, but it's not as redundant - if for some reason the app is closed/refreshed before all requests complete we won't clear anything from disk and they'll be relaunched on next startup
4 - This is also very important and the reason why we use a separate queue - to avoid starting the same requests that we've already started. Because they are still on disk it possible to start them again if we're not careful

---

Without `Onyx.get` I think we can still have a promise based solution that works like this:

```js
Onyx.connect({
    key: ONYXKEYS.NETWORK,
    callback: (network) => {
        // when we transition from offline to online trigger the persisted request queue
        if (isOffline && !network.isOffline) {
            startPersistedRequestsQueue();
        }

        isOffline = network.isOffline;
    },
});

// This is responsible for keeping persistedRequests up to date
Onyx.connect({
    key: ONYXKEYS.NETWORK_REQUEST_QUEUE,
    callback: val => persistedRequests = val || [],
});

function startPersistedRequestsQueue() {
    // If the queue is already running we don't have to do anything
    // it will sort itself in case more requests are persisted after it started
    if (persistedRequestsQueueRunning) {
        return;
    }
    
    persistedRequestsQueueRunning = true;

    // This one time connection replaces `Onyx.get` and signals that data have been read from storage
    const connectionId = Onyx.connect({
        key: ONYXKEYS.NETWORK_REQUEST_QUEUE,
        // We can only process data after it has been read from storage at which point this callback is signalled
        callback: () => {
            Onyx.disconnect(connectionId);
            processPersistedRequestsQueue()
                .finally(() => persistedRequestsQueueRunning = false);
        },
    });

    processPersistedRequestsQueue();
}

function processPersistedRequestsQueue() {
    // This sanity check is also a recursion exit point
    if (_.size(persistedRequests) === 0 || isOffline) {
        return Promise.resolve();
    }

    const tasks = _.map(persistedRequests, request => {
        // processRequest - reusable logic common to both request queues (should return a promise)
        return processRequest(request)
            .then(() => removeFromPersistedStorage(request))  // when a request completes remove it from storage
            .catch((error) => {
                // Decide whether to discard the request despite the error, or keep it for a retry
                if (!shouldRetryRequest(error)) {
                    return removeFromPersistedStorage(request);
                }
            });
    });

    // More request could have been persisted by now
    // If that's the case we'll handle them with a recursive call
    // Otherwise the recursion will exit at the recursion exit point
    return Promise.allSettled(tasks)
        .finally(processPersistedRequestsQueue);
}

// Have something unique in every request and use it to distinct which one we ought to remove
function removeFromPersistedStorage(request) {
    const nextPersistedRequests = _.filter(persistedRequests, r => r.id !== request.id);
    return Onyx.set(ONYXKEYS.NETWORK_REQUEST_QUEUE, nextPersistedRequests);
}
```

**Comment by stephanieelliott** [comment]
I'll hold on this til we agree on an updated strategy and proposal. @johnmlee101 what do you think of the above?

**Comment by johnmlee101** [comment]
I like the proposal! Assigning @kidroca

**Comment by MelvinBot** [comment]
📣 @kidroca You have been assigned to this job by @johnmlee101!
Please apply to this job in Upwork and let us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by kidroca** [comment]
ATM There's no upwork job available for this ticket

I can work on this issue as part of my other ticket: https://github.com/Expensify/App/pull/4908#issuecomment-967139238 because the proposal here: https://github.com/Expensify/App/issues/5987#issuecomment-970510814 would unblock me

**Comment by stephanieelliott** [comment]
Thanks @kidroca, job is on Upwork here: https://www.upwork.com/jobs/~01e307b8bbe03a5385

**Comment by kidroca** [comment]
Hey @stephanieelliott can we reevaluate the price for this ticket as per: https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#payment-for-contributions
> If you think your compensation should be increased for a specific job, you can request a reevaluation by commenting in the Github issue where the Upwork job was posted.

I think this is a bit more technical issue and the current price should be doubled

**Comment by stephanieelliott** [comment]
Hey @johnmlee101, any thoughts on @kidroca's request to double payment for this issue to $500? The CME steps to consider/approve/deny the request are in SO [here](https://stackoverflow.com/c/expensify/questions/10765) 😃

**Comment by johnmlee101** [comment]
Yes! Agreed! This requires a bit more of an overhaul so I'm fine doubling it.

**Comment by stephanieelliott** [comment]
Raised the price to $500, please post a proposal to the job in Upwork when you have a chance @kidroca 

https://www.upwork.com/jobs/~01e307b8bbe03a5385

**Comment by kidroca** [comment]
Proposal submitted,
I expect to have a PR ready by 2021-12-01 EOD

**Comment by kidroca** [comment]
PR is ready

**Comment by MelvinBot** [comment]
@johnmlee101, @kidroca, @stephanieelliott Eep! 4 days overdue now. Issues have feelings too...

**Comment by kidroca** [comment]
Hmm, I don't count Saturday and Sunday Melvin! 😄 

PR still needs review: https://github.com/Expensify/App/pull/6556

**Comment by kidroca** [comment]
PR is open #6556 
I guess we'll see some feedback there soon

**Comment by roryabraham** [comment]
@kidroca I think we should work on nailing down a solid problem statement here so that we can have a richer discussion on solutions here. I think that creating a separate queue for persisted requests is a valid solution, but there are alternatives as well. Since this is very fundamental to the offline-first functionality of our application, I think we'll want to have a wider discussion here.

**Comment by kidroca** [comment]
The problem was originally defined here: https://github.com/Expensify/App/issues/5987#issuecomment-962933224
> "Add Comment" requests are persisted to storage when we're offline
But when we regain connectivity they are read, removed from storage and kept only in memory
Making a refresh/exit before they are complete would make the app "forget" about them

Furthermore we identified another problem with uploading attachments: https://github.com/Expensify/App/pull/4908#issuecomment-967139238
Our current logic that immediately deletes the persisted request (before it's olver) would also deletes the attachment that is yet to be uploaded

---

The problem can be summarized to this
By deleting the persisted request as soon as they starts we:
1. lose the chance to retry it if the app/window gets closed or refreshed before they are over
2. we delete underlying files before they are sent

Clearing a request from persistent storage after it's completed is a robust way to solve the problem

What alternatives do you have in mind?

**Comment by kidroca** [comment]
That's seems good, but I see a few problems:
1. Quite a significant change to the existing queue
2. Terminating the app with `pending` requests
3. `expiresOn` - in this conversation the conclusion is there should be no expiration: https://expensify.slack.com/archives/C01GTK53T8Q/p1638771676235100

(1) I wasn't sure that I can just propose to rewrite the entire queue to work in this matter
(2) If all requests are persisted to disk, then quitting the app with pending requests would trigger them during startup - we only want persist and resend the `shouldPersist` requests.
I guess we can also save the `shouldPersist` flag on the request, and during startup cleanup Onyx of persisted requests that don't have this 

Depending on how the queue works we might be writing way too much data to Onyx - I've brought up cases where we have 30+ requests happening at the same time. 
- that would be more stressful now that it would also involve storage operations
- would the queue still tick every second - more potential disk writes

I'm familiar with the idea to keep everything in Onyx, though I'm it's biggest opponent (I think)
This is a clear example of it's downsides

We would be putting all our request to disk, then 99% of the time remove them, and in between there would be more disk writes to update their state
99% of the time we won't make use of the persisted information
We had performance issues related to too much disk read/writes...

For the 1% of the time where we do need the persisted request - the old solution that uses the `shouldPersist` flag already works 

---

I like you idea, but don't agree with the Onyx part 

We can just as well have a single queue and keep all the requests in local memory `let networkRequestQueue = [];` 
We can still manage them with `pending`, `processing`, `complete` state
We should only persist those that we need to persist - `shouldPersist: true` 
When we clean up requests from `networkRequestQueue` we should check if the request was a `shouldPersist` and remove it from disk as well

But I don't think this is possible with a `setInterval` based queue 
The PR demonstrates a promise based queue that runs requests only when it has to, and can retry them when it has to I think it would be more suitable to your proposal as well

**Comment by kidroca** [comment]
Bump @roryabraham, @johnmlee101 

How should we continue here ?
Overall I agree with @roryabraham's alternative, but it would be like starting this from scratch (If I'm the one working on it at all?)
While what we already have in my PR is decluttering some existing stuff, adds tests and solves the problem as well
I think my changes won't get in the way in a potential refactor that follows Rory's idea so we might as well merge them

**Comment by MelvinBot** [comment]
@johnmlee101, @kidroca, @stephanieelliott Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by johnmlee101** [comment]
Hmm, I'm not quite sure at the moment what the best next step is. @roryabraham do you think we're safe merging the PR then finalizing the discussion about offline states later?

**Comment by roryabraham** [comment]
I think we could, yes. But I just haven't had time to test the PR thoroughly yet. Fortunately the automated tests can give us some additional confidence that things will work as expected.

**Comment by MelvinBot** [comment]
@johnmlee101, @kidroca, @stephanieelliott Eep! 4 days overdue now. Issues have feelings too...

**Comment by johnmlee101** [comment]
No update!

**Comment by stephanieelliott** [comment]
Sounds like we are still testing the PR, not overdue.

**Comment by stephanieelliott** [comment]
Hey @roryabraham are you still working on testing the PR? https://github.com/Expensify/App/pull/6556

**Comment by stephanieelliott** [comment]
^^ Bump @roryabraham, are we still working on testing this?

**Comment by kidroca** [comment]
The PR accidentally got closed, I'll have the chance to open it again and sync with `main` next week
But you leave a comment on the old PR or here if you need anything changed

**Comment by roryabraham** [comment]
Sorry for the delays here, I got behind with the holidays and have been working hard to catch up.

**Comment by johnmlee101** [comment]
Finally got to replying on the PR, sorry about the delays!

**Comment by stephanieelliott** [comment]
PR https://github.com/Expensify/App/pull/6556 was merged.

**Comment by kidroca** [comment]
It might have caused a regression discussed here: https://expensify.slack.com/archives/C01GTK53T8Q/p1644465847015179?thread_ts=1644440925.115879&cid=C01GTK53T8Q

- https://github.com/Expensify/App/issues/7666

We're updating the code with logs and they should be able to tell the full story
- https://github.com/Expensify/App/pull/7673

**Comment by MelvinBot** [comment]
@johnmlee101, @kidroca, @stephanieelliott Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@johnmlee101, @kidroca, @stephanieelliott Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
@johnmlee101, @kidroca, @stephanieelliott Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
@johnmlee101, @kidroca, @stephanieelliott Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@johnmlee101, @kidroca, @stephanieelliott Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@johnmlee101, @kidroca, @stephanieelliott Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
@johnmlee101, @kidroca, @stephanieelliott Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@johnmlee101, @kidroca, @stephanieelliott Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by stephanieelliott** [comment]
Woah, chill Melvin. Not overdue, we are waiting on #6556 to be resolved.

**Comment by stephanieelliott** [comment]
Ah, I see. I think if it's possible to complete this task separately from https://github.com/Expensify/App/issues/7666 then sure, that's fine. @johnmlee101 what do you think?

**Comment by MelvinBot** [comment]
@johnmlee101, @kidroca, @stephanieelliott Huh... This is 4 days overdue. Who can take care of this?

**Comment by johnmlee101** [comment]
Sounds good! Let's do this separately. I have to unassign myself at the moment since I'm no longer part of the CME team.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Gonals (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by kidroca** [comment]
@stephanieelliott, @Gonals
I guess by this you mean the other ticket - https://github.com/Expensify/App/issues/7666

The current task is completed and there's nothing more to be done - no regressions
The other ticket #7666 is tracked separately not part of the current Upwork contract 
_(BTW: It's also resolved and soon to be completed)_

**Comment by MelvinBot** [comment]
Triggered auto assignment to @SofiedeVreese (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @roryabraham (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by stephanieelliott** [comment]
I'm about to go OOO, so reapplied the label to assign this to another CM team member (thanks @SofiedeVreese!)

**Comment by kidroca** [comment]
Can we agree the issue is complete: https://github.com/Expensify/App/issues/5987#issuecomment-1050131124
The PR was merged 22 days ago: https://github.com/Expensify/App/issues/5987#issuecomment-1028958111
And pending in Upwork is due

**Comment by SofiedeVreese** [comment]
@roryabraham just want to get your thumbs up for releasing payment for this GH (5987) which seems to have been actioned via the PR [here](https://github.com/Expensify/App/pull/6556)?

**Comment by roryabraham** [comment]
Yeah, we should release payment here 👍

**Comment by SofiedeVreese** [comment]
Sorry for the long wait there @kidroca I've just released payment.

@Santhosh-Sellavel I'm going to add you to the Upwork job for your reporting bonus now too.

**Comment by MelvinBot** [comment]
Current assignee @rushatgabhane is eligible for the Exported assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @roryabraham is eligible for the Exported assigner, not assigning anyone new.

**Comment by Santhosh-Sellavel** [comment]
Thanks! @SofiedeVreese

**Comment by kidroca** [comment]
Thanks 🎊

**Comment by SofiedeVreese** [comment]
ok reporting bonus paid out too!

All is paid out, removed Upwork posting and closing GH.
